From d6cd959ab4942fdda590796a11ace2ba5e9223b1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Horia=20Geant=C4=83?= <horia.geanta@nxp.com>
Date: Mon, 6 Nov 2017 14:15:20 +0200
Subject: [PATCH 259/767] crypto: testmgr - don't DMA map IV from stack in
 test_skcipher()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Fix the "DMA-API: device driver maps memory from stack" warning
generated when crypto accelerators map the IV.

Note: while so far this has been only a warning and thus a theoretical
issue, on latest kernel trees ablkcipher algorithms fail when IV
is on the stack.
Both caam/jr and caam/qi2 fail in the same way, i.e. same incorrect
ciphertexts. Below is log for caam/qi2:

alg: skcipher: Test 1 failed (invalid result) on encryption for cbc-aes-caam-qi2
00000000: ff 04 80 c3 af 64 47 db 3a 5c b9 b8 91 40 04 8f
alg: skcipher: Test 1 failed (invalid result) on encryption for cbc-3des-caam-qi2
00000000: 22 11 b7 06 29 2a 88 cc 75 46 f4 14 8f d0 44 68
00000010: fe 08 88 30 8d d7 b3 01 21 ca 2d 46 0f 81 2e b6
00000020: ab 2f 38 52 79 00 42 82 5e 31 87 ac 8c 0d 19 88
00000030: 0a 27 5b 4b 6c aa f7 10 80 47 b6 81 83 58 7c 0a
00000040: 63 a5 f7 78 aa 05 11 56 9d 99 06 1d fb 98 01 28
00000050: 8a c0 86 7b 68 62 0c 66 84 b3 31 25 60 ad 13 65
00000060: 54 05 2f cd 4f 9a 1e e3 87 51 dc 11 a7 74 15 c4
00000070: 7a df 68 24 34 50 e6 5e 41 ba 98 48 ca 67 9a 6f
alg: skcipher: Test 1 failed (invalid result) on encryption for cbc-des-caam-qi2
00000000: 52 ee 01 1b 1b a9 59 c7 a6 f5 ea 74 bf 8f f2 2c
00000010: fe d4 df 4a 50 64 3f 2a
alg: skcipher: Test 1 failed (invalid result) on encryption for ctr-aes-caam-qi2
00000000: 4a 3d 28 7a e5 18 b7 84 ce 38 e2 f5 b2 86 58 0b
00000010: a1 04 9f 8f f5 a4 48 bd 0f c6 f3 af 6e d6 cb b3
00000020: 4c f1 52 30 3e bf bd d8 b2 9d 47 a0 40 42 dd 67
00000030: 70 2d 63 f3 1c 9e 96 07 75 29 1e ca db b2 35 ce
alg: skcipher: Test 1 failed (invalid result) on encryption for rfc3686-ctr-aes-caam-qi2
00000000: 26 56 7f bb 8d 68 27 e3 52 e7 0f 51 ee 4f ae 03
alg: skcipher: Test 1 failed (invalid result) on encryption for xts-aes-caam-qi2
00000000: bf 84 80 16 02 00 ec 59 b2 d7 dc 2a 50 22 23 95
00000010: a8 e1 f2 65 15 72 67 1e e3 39 86 6b 4d 27 48 87

Signed-off-by: Horia GeantÄƒ <horia.geanta@nxp.com>
[Xulin: Original patch taken from NXP LSDK-18.09.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 crypto/testmgr.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/crypto/testmgr.c b/crypto/testmgr.c
index 55b978063d0a..e615de1dcb70 100644
--- a/crypto/testmgr.c
+++ b/crypto/testmgr.c
@@ -1312,12 +1312,16 @@ static int __test_skcipher(struct crypto_skcipher *tfm, int enc,
 	struct crypto_wait wait;
 	const char *input, *result;
 	void *data;
-	char iv[MAX_IVLEN];
+	char *iv;
 	char *xbuf[XBUFSIZE];
 	char *xoutbuf[XBUFSIZE];
 	int ret = -ENOMEM;
 	unsigned int ivsize = crypto_skcipher_ivsize(tfm);
 
+	iv = kmalloc(MAX_IVLEN, GFP_KERNEL);
+	if (!iv)
+		return ret;
+
 	if (testmgr_alloc_buf(xbuf))
 		goto out_nobuf;
 
@@ -1538,6 +1542,7 @@ static int __test_skcipher(struct crypto_skcipher *tfm, int enc,
 		testmgr_free_buf(xoutbuf);
 out_nooutbuf:
 	testmgr_free_buf(xbuf);
+	kfree(iv);
 out_nobuf:
 	return ret;
 }
-- 
2.17.0

